### ISSUE TITLE
Recursive object generation

### ISSUE THREAD

**Issue by ObscurusGrassator** [microsoft/TypeScript#43911]
# Bug Report

I need a type check for any object structure with numerical properties with a previously unknown structure.
[stackoverflow]( https://stackoverflow.com/questions/67323006/recursive-object-generation )

[Playground]( https://www.typescriptlang.org/play?ts=4.3.0-dev.20210430#code/C4TwDgpgBAKhDOwoF4oG8DaBrCIBcUiATgJYB2A5gLoFyJQA+UZArgLYBGERAvgNwBYAFDCA9KNgALEvCgyoAQzJQIADwVswAG2gB7AGaLlujgCsIAYyTEWVlkQgA6KAEkoAE13NdSLGV0A7nLAssBELNA2dg5yygruAG5KFk7COkjACMC0WSjoCgRoHIUWeABMPAA0HgRl1foKJFruBADkXBYKLPCRYeQUrTz8UOJQoJBQFpKWWP1QAbpEWLIUurruwmIS49BTM3MAcgDyMPOLy1D6i4paWlAGY9OXureBc2oa2ghQALRQAEIAVQA4lAAPxbMZZRyqRwmUwoAB8RhAkIACkRdJAiKAoK14a0PLpvv4kGoZEhdModnjWJxuIxYFlWo5IVAoBisdxcfizITPCSfCpVBT7tTwNBWnSuEQWQAKMoAZkVAE4AJTCTKIGFwsyOECOABegiEWuAOvh+qNeTKJrNFr1BoaTXceXali6PUIfUorTt0Nh8P92sDZhtwfNodMjmdzTdHU9vVIvpNQA )

### 💻 Code

```ts
type Test = {[key: string]: Test | number};

// This is an example of an object structure. I do not know its true structure in advance.
let test: Test = {a: {b: {c:2}, d: 2, faild: 'because string'}}; // type checking works good

// type checking NOT works for all of the following examples - BUG ?
// test.x.obj => any
// Property 'obj' does not exist on type 'number | Test'.
//   Property 'obj' does not exist on type 'number'.(2339)
test.x.obj.y.z;
test.x.obj.y.z = 2;
test.x.obj.y.faild = 'because string';
test.x.obj;
test.x.obj = 2;
test.x.obj.faild = 'because string';
```

**Comment by RyanCavanaugh** [comment]
It looks like this is a question rather than a bug report. This issue tracker is for tracking bugs and active work on TypeScript *itself*, rather than a general forum for programmers using TypeScript to get help or ask questions.

You can ask questions on sites like [Stack Overflow](https://stackoverflow.com/questions/tagged/typescript). We are not able to provide one-on-one support on the issue tracker. Please read the issue template carefully - it has important information on what kinds of reports can be acted on here, as well as links to useful TypeScript resources. Thanks!

**Comment by ObscurusGrassator** [comment]
Before creating this bug, I created a question on Stackoverflow (the link is attached in bug), so that I don't burden you unnecessarily. Before that, I consulted with colleagues at work and looked for possible solutions via documentation and google. I did create this bug after I not received a relevant response. Are you saying the Typescript behavior is correct? I apologize if I created this Bug request incorrectly.
